File: rust/parity/specs/data_retention_policy_enforcer.py
"""Parity spec for PiDataRetentionPolicyEnforcer.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiDataRetentionPolicyEnforcer"

_mod = load_py_agent("pi_data_retention_policy_enforcer.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiDataRetentionPolicyEnforcer()
    out = agent.enforce_retention(_mod.RetentionInput(**data))
    return out.model_dump()


_CLEAN = "retention: 30d\ndelete: scheduled\npurge: enabled\nconsent_check: true"
_INDEFINITE = "retain: indefinite\nlifecycle: none"
_DELETE_NEVER = "policy:\n  delete: never\n  archive: cold-storage"
_UNLIMITED = "RETENTION: UNLIMITED\nbackups: kept"
_PII_NO_CONSENT = "pii: retain\nconsent_check: false"
_PERSONAL_SAVE = "personal_data: save\nconsent: false"
_PII_WITH_CONSENT = "pii: retain\nconsent_check: true"

SAMPLES = [
    # clean / passing policy
    {"input": {"policy_content": _CLEAN}},
    # indefinite retention -> risk 80, fails in strict mode
    {"input": {"policy_content": _INDEFINITE}},
    # delete: never variant
    {"input": {"policy_content": _DELETE_NEVER}},
    # retention: unlimited variant (uppercase, exercises .lower())
    {"input": {"policy_content": _UNLIMITED}},
    # PII retained without consent -> risk 90
    {"input": {"policy_content": _PII_NO_CONSENT}},
    # personal_data save without consent -> risk 90
    {"input": {"policy_content": _PERSONAL_SAVE}},
    # PII present but consent check passes -> no issue, secure
    {"input": {"policy_content": _PII_WITH_CONSENT}},
    # empty / edge input
    {"input": {"policy_content": ""}},
    # non-strict env -> WARN path, is_secure stays True despite risk
    {"input": {"policy_content": _INDEFINITE},
     "env": {"PI_RETENTION_STRICT_MODE": "false"}},
    # explicit strict env -> FAILED_COMPLIANCE
    {"input": {"policy_content": _INDEFINITE},
     "env": {"PI_RETENTION_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/data_retention_policy_enforcer.rs
//! Port of `pi_micro_agents/pi_data_retention_policy_enforcer.py`.
//!
//! Verifies automated data deletion schedules, purging PII records, and
//! enforcing minimal storage lifetimes. Behaviour is a line-for-line mirror of
//! the Python original.

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub policy_content: String,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub issues: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: strict unless the env var is set, in which case
/// it is strict only when the value equals (case-insensitively) "true".
fn is_strict_mode() -> bool {
    match std::env::var("PI_RETENTION_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn enforce_retention(input: &Input) -> Output {
    let content = input.policy_content.to_lowercase();
    let mut issues: Vec<String> = Vec::new();
    let mut risk_score: f64 = 0.0;

    // Retaining data indefinitely
    if content.contains("retain: indefinite")
        || content.contains("delete: never")
        || content.contains("retention: unlimited")
    {
        issues.push(
            "Indefinite Data Retention: Configuration stores user records indefinitely without automated purge triggers."
                .to_string(),
        );
        risk_score = risk_score.max(80.0);
    }

    // Retention of PII without strict consent controls
    if content.contains("pii: retain") || content.contains("personal_data: save") {
        if content.contains("consent_check: false") || content.contains("consent: false") {
            issues.push(
                "Uncontrolled PII Retention: Sensitive personal identifiers stored without mandatory consent checks."
                    .to_string(),
            );
            risk_score = risk_score.max(90.0);
        }
    }

    let mut is_sec = true;
    if risk_score > 30.0 && is_strict_mode() {
        is_sec = false;
    }

    let mut status = if is_sec {
        "PASSED".to_string()
    } else {
        "FAILED_COMPLIANCE".to_string()
    };
    if risk_score > 0.0 && is_sec {
        status = "WARN_COMPLIANCE".to_string();
    }

    Output {
        is_secure: is_sec,
        issues,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = enforce_retention(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(content: &str) -> Output {
        enforce_retention(&Input {
            policy_content: content.into(),
        })
    }

    #[test]
    fn clean_policy_passes() {
        let o = run("retention: 30d\ndelete: scheduled\npurge: enabled");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.issues.is_empty());
    }

    #[test]
    fn indefinite_retention_flagged() {
        let o = run("retain: indefinite");
        assert!(!o.is_secure);
        assert_eq!(o.status, "FAILED_COMPLIANCE");
        assert_eq!(o.risk_score, 80.0);
        assert_eq!(o.issues.len(), 1);
    }

    #[test]
    fn pii_without_consent_flagged() {
        let o = run("pii: retain\nconsent: false");
        assert!(!o.is_secure);
        assert_eq!(o.status, "FAILED_COMPLIANCE");
        assert_eq!(o.risk_score, 90.0);
    }

    #[test]
    fn pii_with_consent_not_flagged() {
        // PII retention present but consent check is not false -> no issue.
        let o = run("pii: retain\nconsent_check: true");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
    }
}
